feat(om2): add native histograms to OpenMetrics2.0 (#2634)
prometheus/docs: docs/specs/om/open_metrics_spec_2_0.md

#### Values

Metric values in OpenMetrics MUST be either numbers or complex data types.

Numbers MUST be either floating points or integers. Note that ingestors of the format MAY only support float64. The non-real values NaN, +Inf and -Inf MUST be supported. NaN MUST NOT be considered a missing value, but it MAY be used to signal a division by zero.

Complex data types MUST contain all information necessary to recreate a sample of a Metric Type, with the exception of Created Value and Exemplars.

List of complex data types:
- Integer counter native histograms for the Metric Type Histogram.

> - Integer gauge native histograms for the Metric Type GaugeHistogram.

Author: Haven't thought about it. There's two options: allow floats everywhere and use absolute numbers for buckets similar to how we do regular float counters. Which means we'd have a single exposition format for both.

Or introduce a new kind of exposition to be detected from the fields that occur (deltas vs buckets).

> - Integer gauge native histograms for the Metric Type GaugeHistogram.

Author: I have a suggestion to use `fcount`, `fsum`, `fzero_count` instead of the original and also `positive_buckets` instead of `positive_deltas` when exposing floats. This mimics the protobuf and makes it easy to decide early what we're dealing with. Also makes it possible to move this question to a next PR :)

> - Integer gauge native histograms for the Metric Type GaugeHistogram.

Author: All things considered, I think we should make the decision based on what kind of number `count` field has. And by kind of number I mean: integer of floating point based on  https://prometheus.io/docs/specs/om/open_metrics_spec/#numbers .

#### Values

Metric values in OpenMetrics MUST be either numbers or complex data types.

Numbers MUST be either floating points or integers. Note that ingestors of the format MAY only support float64. The non-real values NaN, +Inf and -Inf MUST be supported. NaN MUST NOT be considered a missing value, but it MAY be used to signal a division by zero.

> Complex data types MUST contain all information necessary to recreate a sample of a Metric Type, with the exception of Created Timestamp and Exemplars.

Reviewer: ```suggestion
Complex data types MUST contain all information necessary to recreate a sample of a Metric Type, with the exception of Start Timestamp and Exemplars.
```

Reviewer: Assuming this merges after the created -> start PR

Author: It will be easier to do it after the other PR is merged actually, deferring.

#### Values

Metric values in OpenMetrics MUST be either Number or ComplexValue.

##### Number

> Number value MUST be either floating point or integer. Note that ingestors of the format MAY only support float64. The non-real values NaN, +Inf and -Inf MUST be supported. NaN value MUST NOT be considered a missing value, but it MAY be used to signal a division by zero.

Author: @bwplotka I've removed the forward reference to the format and made "Number" and "ComplexValue" it's own section to be able to refer to it if we want and follow the convention.

Histograms measure distributions of discrete events. Common examples are the latency of HTTP requests, function runtimes, or I/O request sizes.

A Histogram MetricPoint MUST contain Count, Sum values.

The Count value MUST be equal to the number of measurements taken by the Histogram. The Count is a counter semantically. The Count MUST be an integer and MUST NOT be NaN or negative.

The Sum value MUST be equal to the Sum of all the measured event values. The Sum is only a counter semantically as long as there are no negative event values measured by the Histogram MetricPoint.

> A Histogram MetricPoint MUST contain either [classic buckets](#classic-buckets) or [exponential buckets](#exponential-buckets) or both.

Reviewer 2: I think we should call the "exponential buckets" better "native buckets". See comment below for more rationale.

In different news, the "exponential buckets" section later states that it is valid to have no exponential buckets. Down there, it sounds like that would mean just the zero bucket. However, the zero bucket is also marked as optional further down. That's a bit confusing. In protobuf, the zero bucket is always there (in doubt as a bucket of width zero and with zero population). This could be seen as the "default" if no zero bucket is specified, so it would be fine to not enforce a zero bucket in the OMv2 text format. However, protobuf also allows an "empty" histograms (which then still has the zero bucket as described above). For the text format, that would mean we need to allow an empty histogram, too, with count and sum equals zero, which is not obvious from the current wording. Another special and extreme case is a native histograms (without any classic buckets) that has only observations of NaN. It would have a count > 0, a sum == NaN, and no buckets whatsoever (beyond the zero bucket, which is optional here).

(OMv1 apparently wanted to not have bucket-less histograms because that would be a summary without quantiles. However, even unpopulated classic buckets are represented, but that's not the case with native histograms.) 

Not sure how to phrase this. I think we need to drop the "MUST have any buckets", but we can say that any non-NaN observations MUST go to buckets (either classic or native).

Every bucket MUST have well defined boundaries and a value. Boundaries of a bucket MUST NOT be NaN. Bucket values MUST be integers. Semantically, bucket values are counters so MUST NOT be NaN or negative.

A Histogram SHOULD refuse to measure NaN value as adding NaN to the Sum will make the Sum equal to NaN and mask the sum of the real measurements until the next reset of the counters. If a Histogram does allow NaN, then NaN MUST be counted in the Count and MUST be added to the Sum, resulting in the Sum becoming NaN.

A Histogram MAY refuse to measure +Inf and -Inf values as adding these to the Sum will mask the sum of the real measurements until the next reset of the counters. If a Histogram measures +Inf or -Inf, then +Inf or -Inf MUST be counted in the Count and MUST be added to the Sum, potentially resulting in +Inf, -Inf or NaN in the Sum, the later for example in case of adding +Inf to -Inf.

A Histogram MetricPoint SHOULD have a Timestamp value called Created. This can help ingestors discern between new metrics and long-running ones it did not see before.

A Histogram's Metric's LabelSet MUST NOT have a "le" label name.

> ##### Classic buckets

Reviewer 2: Somehow I'm missing that classic buckets should aim for the same bucket layout over time (e.g. even unpopulated buckets should be represented) while native buckets are even encouraged to not be represented if they are empty.

Author: I don't think we can require that the layout doesn't change.
I think exposing unpopulated buckets is a matter of the exposition , not the data model, so I'm making a change there. cbfe910edac9b4050712534e8226b52d7ae11d11

Author: For native histograms: 16b4c50467e715e68e1a498804929343c7ce41dc and ca1d860f84b92a06e1d9e28dee40d65d81b38d5a

As an example for a metric representing request latency in seconds its values for classic buckets with thresholds 1, 2, 3, and +Inf MUST follow value_1 <= value_2 <= value_3 <= value_+Inf. If ten requests took 1 second each, the values of the 1, 2, 3, and +Inf buckets MUST equal 10.

Histogram MetricPoints with classic buckets MUST have one classic bucket with a +Inf threshold. The +Inf bucket counts all requests.

If the NaN value is not allowed, then the Count value MUST be equal to the value of the +Inf bucket.

If the NaN value is allowed, it SHOULD be counted in the +Inf bucket, and MUST not be counted in any other bucket. In case the NaN is counted in the +Inf bucket, then the Count MUST be equal to the value of the +Inf bucket, otherwise the Count MUST be greater. The ratonale is that NaN does not belong to any bucket mathematically, however instrumentation libraries traditionally put it into the +Inf bucket, which is why the wording "SHOULD" is used.

Negative threshold classic buckets MAY be used.

> Classic bucket values MAY have exemplars. The value of the exemplar MUST be within the classic bucket. Exemplars SHOULD be put into the classic bucket with the highest threshold. A classic bucket MUST NOT have more than one exemplar.

Author: fixed, I think the intent was the same

##### Exponential buckets

> Histogram MetricPoints with exponential buckets MUST have a Schema value. The Schema is an 8 bit signed integer between -4 and 127. Schema values between -4 and 127 are also called Standard Schemas.

Author: 52bd9e9411c22825bd96b25d153d14b5c10859e3

> For any Standard Schema n, the Histogram MetricPoint MAY contain positive, negative exponential buckets and a single zero bucket. It is valid to have no exponentual buckets at all.

Reviewer 2: What also strikes me as weird (in general) that the whole section is called "exponential buckets" but then a "zero bucket" is included that is in fact not an "exponential bucket". 

Furthermore, we might one day add schemas that are not exponential (but aren't the special case of NHCB, either), and then we needed to rename this section.

Maybe call this section "Native buckets"?

> For any Standard Schema n, the Histogram MetricPoint MAY contain positive, negative exponential buckets and a single zero bucket. It is valid to have no exponentual buckets at all.

Author: renamed to native buckets everywhere and now refer to zero bucket as zero native bucket.

The positive exponential bucket that contains MaxFloat64 (according to the boundary formulas above) has an upper inclusive limit of MaxFloat64 (rather than the limit calculated by the formulas above, which would overflow float64).

The next positive exponential bucket (index i+1 relative to the bucket from the previous item) has a lower exclusive limit of MaxFloat64 and an upper inclusive limit of +Inf. (It could be called a positive exponential overflow bucket.)

The negative exponential bucket that contains MinFloat64 (according to the boundary formulas above) has a lower inclusive limit of MinFloat64 (rather than the limit calculated by the formulas above, which would underflow float64).

The next negative exponential bucket (index i+1 relative to the bucket from the previous item) has an upper exclusive limit of MinFloat64 and an lower inclusive limit of -Inf. (It could be called a negative exponential overflow bucket.)

Exponential buckets beyond the +Inf and -Inf buckets described above MUST NOT be used.

> If the zero bucket is present, the Historam MetricPoint MUST have a Zero threshold. The Zero threshold MUST be a non-negative float64 value (threshold >= 0.0). The boundaries of the Zero native bucket are `[-threshold, threshold]` inclusive.

Reviewer 2: Maybe "inclusively" is better grammatically?

Author: done

> If the zero bucket is present, any measured value that falls into the zero bucket MUST be counted towards the zero bucket and MUST NOT be counted in any other exponential bucket. The Zero threshold SHOULD be equal to a lower limit of an arbitraty exponential bucket.

Reviewer 2: arbitraty → arbitrary

Author: done

If the NaN value is not allowed, then the Count value MUST be equal to the sum of the negative, positive and zero buckets.

> If the NaN value is allowed, it SHOULD NOT be counted in any bucket. The value NaN MAY be counted in the +Inf bucket, and MUST not be counted in any other bucket. In case the NaN is counted in the +Inf bucket, then the Count value MUST be equal to the sum of the negative, positive and zero buckets, otherwise the Count MUST be greater. The ratonale is that NaN does not belong to any bucket mathematically, and instrumentation libraries traditionally don't put it into any bucket, which is why the wording "SHOULD NOT" is used.

Reviewer 2: For exponential buckets, I would say NaN MUST NOT be counted in any bucket.

Author: yes, I think you're especially since we went to great lengths to detect and warn based on this fact https://github.com/prometheus/prometheus/pull/16724

A Histogram MetricPoint with exponential buckets MAY contain exemplars.

Exemplars associated with a Histogram MetricPoint with exponential buckets MUST have a timestamp.

The values of exemplars in a Histogram MetricPoint with exponential buckets MUST fall into one of the exponential buckets.

The values of exemplars in a Histogram MetricPoint with exponential buckets SHOULD be evenly distributed to avoid only representing the bucket with the highest value and therefore most common case.

#### GaugeHistogram

GaugeHistograms measure current distributions. Common examples are how long items have been waiting in a queue, or size of the requests in a queue.

> A GaugeHistogram MetricPoint MUST contain either classic buckets or exponential buckets or both.

Reviewer 2: Same bucket requirement concerns as above.

A GaugeHistogram MetricPoint SHOULD contain Gcount, Gsum. Every bucket MUST have well defined boundaries and a value. Boundaries of a bucket MUST NOT be NaN. Gcount and bucket values MUST be integers.

> The bucket and Gsum of a GaugeHistogram are conceptually gauges, however bucket values MUST NOT be negative or NaN. If negative threshold buckets are present, then Gsum MAY be negative. Gsum MUST NOT be NaN. Bucket values MUST be integers.

Reviewer 2: I think we should drop "If negative threshold buckets are present, then Gsum MAY be negative." This makes classic histograms without negative bucket threshold but negative observations unrepresentable for no reason.

"Gsum MUST NOT be NaN." is also problematic because this is the case if NaN observations are in the histogram, and therefore those histograms are unrepresantable.

> A GaugeHistogram's Metric's LabelSet MUST NOT have a "le" label name.

Reviewer 2: Unless it's only native buckets.

Author: done in 9d02b5b8f28cc84b9af8937532f2eec3804defd3

> The buckets for a GaugeHistogram follow all the same rules as for a Histogram, with Gcount playing the same role as Count.

Reviewer 2: And Gsum as Sum?

If the zero native bucket is present, the Historam MetricPoint MUST have a Zero threshold. The Zero threshold MUST be a non-negative float64 value (threshold >= 0.0). The boundaries of the zero native bucket are `[-threshold, threshold]` inclusively.

If the zero native bucket is present, any measured value that falls into the zero native bucket MUST be counted towards the zero native bucket and MUST NOT be counted in any other native bucket. The Zero threshold SHOULD be equal to a lower limit of an arbitrary native bucket.

If the NaN value is not allowed, then the Count value MUST be equal to the sum of the negative, positive and zero native buckets.

If the NaN value is allowed, it MUST NOT be counted in any bucket and the Count MUST be greater than the sum of the negative, positive and zero native buckets. The ratonale is that NaN does not belong to any bucket mathematically, and instrumentation libraries traditionally don't put it into any bucket.

A Histogram MetricPoint with native buckets MAY contain exemplars.

Exemplars associated with a Histogram MetricPoint with native buckets MUST have a timestamp.

> The values of exemplars in a Histogram MetricPoint with native buckets MUST fall into one of the native buckets.
> 

Author: I think I'll remove this requirement, since:
1) exemplar value will be a real number so by definition some native bucket will cover it
2) after reset the actual stored bucket may be missing, but the exemplar can be there.

```suggestion
```

> The values of exemplars in a Histogram MetricPoint with native buckets MUST fall into one of the native buckets.
> 

Author: I think I'll remove this, since:
1) if the exemplar's value is not NaN then definition some native bucket will cover it
2) if the exemplar's value is NaN this is not true
3) after reset the actual stored bucket may be missing, but the exemplar can be there.

```suggestion
```

Histograms measure distributions of discrete events. Common examples are the latency of HTTP requests, function runtimes, or I/O request sizes.

A Histogram MetricPoint MUST contain Count, Sum values.

The Count value MUST be equal to the number of measurements taken by the Histogram. The Count is a counter semantically. The Count MUST be an integer and MUST NOT be NaN or negative.

The Sum value MUST be equal to the Sum of all the measured event values. The Sum is only a counter semantically as long as there are no negative event values measured by the Histogram MetricPoint.

> A Histogram MUST measure non NaN values in either [classic buckets](#classic-buckets) or [native buckets](#native-buckets) or both. If a Histogram stops measuring values in either classic or native buckets and keeps measuring values in the other, it MUST clear and not expose the buckets it stopped measuring into. This avoids exposing different distribution from the two kind of buckets at the same time.

Reviewer 2: "non NaN values" → "non-NaN values" or "values that are not NaN" (currently, I find this hard to read)

Author: going with second option

If the Histogram Metric has MetricPoints with classic buckets, the Histogram's Metric's LabelSet MUST NOT have a "le" label name.

##### Classic buckets

Every classic bucket MUST have a threshold. Classic bucket thresholds within a MetricPoint MUST be unique.

A classic bucket MUST cover every measured value less or equal to its threshold, or to put it another way, classic buckets MUST be cumulative. Classic buckets are cumulative to allow monitoring systems to drop any non-+Inf bucket for performance/anti-denial-of-service reasons in a way that loses granularity but is still a valid Histogram.

As an example for a metric representing request latency in seconds its values for classic buckets with thresholds 1, 2, 3, and +Inf MUST follow value_1 <= value_2 <= value_3 <= value_+Inf. If ten requests took 1 second each, the values of the 1, 2, 3, and +Inf buckets MUST equal 10.

Histogram MetricPoints with classic buckets MUST have one classic bucket with a +Inf threshold. The +Inf bucket counts all requests.

The Count value MUST be equal to the value of the +Inf bucket.

> If the NaN value is allowed, it MUST be counted in the +Inf bucket, and MUST not be counted in any other bucket. The ratonale is that NaN does not belong to any bucket mathematically, however instrumentation libraries traditionally put it into the +Inf bucket.

Reviewer 2: ```suggestion
If the NaN value is allowed, it MUST be counted in the +Inf bucket, and MUST not be counted in any other bucket. The rationale is that NaN does not belong to any bucket mathematically, however instrumentation libraries traditionally put it into the +Inf bucket.
```

The negative native bucket that contains MinFloat64 (according to the boundary formulas above) has a lower inclusive limit of MinFloat64 (rather than the limit calculated by the formulas above, which would underflow float64).

The next negative native bucket (index i+1 relative to the bucket from the previous item) has an upper exclusive limit of MinFloat64 and an lower inclusive limit of -Inf. (It could be called a negative native overflow bucket.)

Native buckets beyond the +Inf and -Inf buckets described above MUST NOT be used.

If the zero native bucket is present, the Historam MetricPoint MUST have a Zero threshold. The Zero threshold MUST be a non-negative float64 value (threshold >= 0.0). The boundaries of the zero native bucket are `[-threshold, threshold]` inclusively.

If the zero native bucket is present, any measured value that falls into the zero native bucket MUST be counted towards the zero native bucket and MUST NOT be counted in any other native bucket. The Zero threshold SHOULD be equal to a lower limit of an arbitrary native bucket.

If the NaN value is not allowed, then the Count value MUST be equal to the sum of the negative, positive and zero native buckets.

> If the NaN value is allowed, it MUST NOT be counted in any bucket and the Count MUST be greater than the sum of the negative, positive and zero native buckets. The ratonale is that NaN does not belong to any bucket mathematically, and instrumentation libraries traditionally don't put it into any bucket.

Reviewer 2: ```suggestion
If the NaN value is allowed, it MUST NOT be counted in any bucket and the Count MUST be greater than the sum of the negative, positive and zero native buckets. The rationale is that NaN does not belong to any bucket mathematically, and instrumentation libraries traditionally don't put it into any bucket.
```

Beyond typos: Not sure if we can talk about a tradition already. Not counting NaN observations in any bucket was part of the nascent spec pretty early. So it didn't just happen out of tradition.

"Count MUST be greater than the sum..." → that's only the case if there are NaN observations in the histogram. Maybe use a wording that implies that the difference between Count and the sum of all buckets MUST be equal to the number of NaN observations.

A Histogram MUST measure values that are not NaN in either [Classic Buckets](#classic-buckets) or [Native Buckets](#native-buckets) or both. Measuring NaN is different for Classic and Native Buckets, see in their respective sections.

If a Histogram stops measuring values in either Classic or Native Buckets and keeps measuring values in the other, it MUST clear and not expose the buckets it stopped measuring into. This avoids exposing different distribution from the two kind of buckets at the same time.

Every Bucket MUST have well defined boundaries and a value. Boundaries of a Bucket MUST NOT be NaN. Bucket values MUST be integers. Semantically, bucket values are counters so MUST NOT be NaN or negative.

A Histogram SHOULD refuse to measure NaN value as adding NaN to the Sum will make the Sum equal to NaN and mask the sum of the real measurements until the next reset of the counters. If a Histogram does allow NaN, then NaN MUST be counted in the Count and MUST be added to the Sum, resulting in the Sum becoming NaN.

> A Histogram MAY refuse to measure +Inf and -Inf values as adding these to the Sum will mask the sum of the real measurements until the next reset of the counters. If a Histogram measures +Inf or -Inf, then +Inf or -Inf MUST be counted in the Count and MUST be added to the Sum, potentially resulting in +Inf, -Inf or NaN in the Sum, the later for example in case of adding +Inf to -Inf.

Author: Not really. I can reword to say : if +-Inf is included, then what to do and add an explanation what it means. So basically flip the two sentences around and not use "MAY" . Something like:
"If a Histogram includes +Inf or -Inf measurement, then +Inf or -Inf MUST be counted in Count and MUST be added to the Sum, potentially resulting in +Inf, -Inf or NaN in the Sum, the later for example in case of adding +Inf to -Inf. Note that in this case the Sum of finite measurements is masked until the next reset of the Histogram."

WDYT?

Native Buckets beyond the +Inf and -Inf buckets described above MUST NOT be used.

The boundaries of the zero Native Bucket are `[-threshold, threshold]` inclusively. The Zero threshold MUST be a non-negative float64 value (threshold >= 0.0).

If the Zero threshold is positive (threshold > 0), then any measured value that falls into the zero Native Bucket MUST be counted towards the zero Native Bucket and MUST NOT be counted in any other native bucket. The Zero threshold SHOULD be equal to a lower limit of an arbitrary Native Bucket.

If the NaN value is not allowed, then the Count value MUST be equal to the sum of the negative, positive and zero Native Buckets.

If the NaN value is allowed, it MUST NOT be counted in any Native Bucket, and MUST be counted towards the Count. The difference between the Count and the sum of the negative, positive and zero Native Buckets MUST BE the number of NaN observations. The rationale is that NaN does not belong to any bucket mathematically.

A Histogram MetricPoint with Native Buckets MAY contain exemplars.

> Exemplars associated with a Histogram MetricPoint with Native Buckets SHOULD have a timestamp. Note: storage implementations may drop exemplars without timestamps if keeping track of exemplars without timestamps is too resource intensive.

Reviewer: > I think saying MUST is out of the question.

FYI this is what OTel libraries do. Recording a span requires taking the current timestamp twice (start + end), so if exemplars are used for sampled spans (which is usually the case), the overhead isn't bad, and is controllable via sampling rate.  I'm not sure for other use-cases, though.  OTel libraries also use a time-weighted algorithm for storing exemplars (instead of always keeping the last one), which can further reduce the need to take the current time.

I could also imagine ways to implement lower-resolution timestamp additions as a way to mitigate this... E.g. store the "roughly current" timestamp as a global atomic variable that can be used by recorded exemplars.  Have a goroutine update the "roughly current" timestamp once-per-second.

Seems like our viable options are:

* Require it at the protocol level (assuming you buy my reasoning above).
* Require that exemplars exposed without a timestamp only be returned by a single scrape (i.e. be deleted from storage on scrape).

I think the splitting of exemplars between messages isn't too big of a deal.  They are essentially just events attached to the metric.

We will need to solve the multiple exemplars per data-point problem for native histograms and for "complex" classic histograms and summaries...  So back to the original suggestion, I think we can just add this statement to the general exemplar spec section, and discuss these changes separately.  Saying SHOULD for timestamps on exemplars is non-controversial, right?

Histogram MetricPoints with Classic Buckets MUST have one Classic Bucket with a +Inf threshold. The +Inf bucket counts all measurements.

Exposed Classic Bucket thresholds SHOULD stay constant over time and between targets whose metrics are intended to be aggregated. A change of thresholds may prevent the affected histograms to be part of the same operation (e.g. an aggregation of different metrics or a rate calculation over time).

The Count value MUST be equal to the value of the +Inf bucket.

If the NaN value is allowed, it MUST be counted in the +Inf bucket, and MUST not be counted in any other bucket. The rationale is that NaN does not belong to any bucket mathematically, however instrumentation libraries traditionally put it into the +Inf bucket.

Classic Bucket values MAY have exemplars. The value of the exemplar MUST be within the Classic Bucket. Exemplars SHOULD be put into the Classic Bucket with the lowest threshold that includes the exemplar value. A Classic Bucket MUST NOT have more than one exemplar.

##### Native Buckets

> Histogram MetricPoints with Native Buckets MUST have a Schema value. The Schema is an 8 bit signed integer between -4 and 8. Schemas between -9 and 52 are called Standard (exponential) Schemas, the currently unused Schemas -9 to -5 and 9 to 52 are reserved to be used as Standard Schemas later.

Reviewer 3: ```suggestion
Histogram MetricPoints with Native Buckets MUST have a Schema value. The Schema is an 8 bit signed integer between -4 and 8. 

* Schema values between -9 and 52 are called Standard (exponential) Schemas
* Schema values between -9 to -5 and 9 to 52 are reserved to be used as Standard Schemas later.
* Schema value equal to -53 is reserved.
```

Histograms measure distributions of discrete events. Common examples are the latency of HTTP requests, function runtimes, or I/O request sizes.

A Histogram MetricPoint MUST contain Count and Sum values.

The Count value MUST be equal to the number of measurements taken by the Histogram. The Count is a counter semantically. The Count MUST be an integer and MUST NOT be NaN or negative.

The Sum value MUST be equal to the sum of all the measured event values. The Sum is only a counter semantically as long as there are no negative event values measured by the Histogram MetricPoint.

A Histogram MUST measure values that are not NaN in either [Classic Buckets](#classic-buckets) or [Native Buckets](#native-buckets) or both. Measuring NaN is different for Classic and Native Buckets, see in their respective sections.

> Every Bucket MUST have well defined boundaries and a value. The bucket value is called the bucket count colloquially. Boundaries of a Bucket MUST NOT be NaN. Bucket values MUST be integers. Semantically, bucket values are counters so MUST NOT be NaN or negative.

Author: yes, but there was no issue: https://github.com/prometheus/OpenMetrics/issues/309

A Histogram SHOULD NOT include NaN measurements as including NaN in the Sum will make the Sum equal to NaN and mask the sum of the real measurements for the lifetime of the time series. If a Histogram includes NaN measurements, then NaN measurements MUST be counted in the Count and the Sum MUST be NaN.

If a Histogram includes +Inf or -Inf measurement, then +Inf or -Inf MUST be counted in Count and MUST be added to the Sum, potentially resulting in +Inf, -Inf or NaN in the Sum, the later for example in case of adding +Inf to -Inf. Note that in this case the Sum of finite measurements is masked until the next reset of the Histogram.

A Histogram MetricPoint SHOULD have a Timestamp value called Start Timestamp. This can help ingestors discern between new metrics and long-running ones it did not see before.

If the Histogram Metric has MetricPoints with Classic Buckets, the Histogram's Metric's LabelSet MUST NOT have a "le" label name.

##### Classic Buckets

> Every Classic Bucket MUST have a threshold. Classic Bucket thresholds within a MetricPoint MUST be unique. Classic Buckets MAY have a negative threshold.

Author: 0b46a71997c2013f414dae310cbe7a0ab16c8148

Histogram MetricPoints with Classic Buckets MUST have one Classic Bucket with a +Inf threshold. The +Inf bucket counts all measurements.

Exposed Classic Bucket thresholds SHOULD stay constant over time and between targets whose metrics are intended to be aggregated. A change of thresholds may prevent the affected histograms to be part of the same operation (e.g. an aggregation of different metrics or a rate calculation over time).

The Count value MUST be equal to the value of the +Inf bucket.

If the NaN value is allowed, it MUST be counted in the +Inf bucket, and MUST not be counted in any other bucket. The rationale is that NaN does not belong to any bucket mathematically, however instrumentation libraries traditionally put it into the +Inf bucket.

Classic Bucket values MAY have exemplars. The value of the exemplar MUST be within the Classic Bucket. Exemplars SHOULD be put into the Classic Bucket with the lowest threshold that includes the exemplar value. A Classic Bucket MUST NOT have more than one exemplar.

##### Native Buckets

Histogram MetricPoints with Native Buckets MUST have a Schema value. The Schema is an 8 bit signed integer between -4 and 8. 

> * Schema values between -9 and 52 are called Standard (exponential) Schemas

Author: hmm, maybe we should just not have this list and say that:
```
The Schema MUST be an 8 bit signed integer between -4 and 8 (inclusive).
```

We might add that "Other values MUST NOT be used." , although that's redundant.

> * Schema values between -9 to -5 and 9 to 52 are reserved to be used as Standard Schemas later.

Author: I've reworded this in 5eb619797bb50d345cb405ef4ee3e3f63b1e3321

I don't think we need to say anything about compatibility or what to do with unknown schemas. The standard is clear, unknown (to the ingestor) schemas are not valid and will be rejected. We can add support for more schemas by bumping the minor version in the spec.

> * Schema values between -9 to -5 and 9 to 52 are reserved to be used as Standard Schemas later.

Author: We could say that the upper limit for standard schemas is 52 and allow that already, since implementations can easily downscale to 8 if they don't want to store such high resolution. hmm?

* Schema value equal to -53 is reserved.

For any Standard Schema n, the Histogram MetricPoint MAY contain positive, negative Native Buckets and MUST contain a zero Native Bucket. Empty positive or negative Native Buckets SHOULD NOT be present.

> As the Standard Schema allows for many positive and negative Native Buckets, the Histogram SHOULD have strategies to reset itself to empty when the number of Native Buckets is too high. Such strategies are out of scope for this specification.

Reviewer: Perfect

metric-type = counter / gauge / histogram / gaugehistogram / stateset
metric-type =/ info / summary / unknown

> sample = metricname [labels] SP value [SP timestamp] [exemplar] LF

Author: todo, allow multiple exemplars for native histograms

metric-type =/ info / summary / unknown

sample = metricname-and-labels SP number [SP timestamp] [SP start-timestamp] [exemplar] LF

> sample =/ metricname-and-labels SP "{" complextype "}" [SP timestamp] [SP start-timestamp] *exemplar LF

Author: ack

Quantiles MAY be in any order.

##### Histogram with classic buckets

The MetricPoint's Classic Bucket Values Sample MetricNames MUST have the suffix `_bucket`. If present, the MetricPoint's Sum Value Sample MetricName MUST have the suffix `_sum`. If present, the MetricPoint's Created Value Sample MetricName MUST have the suffix `_created`.
If and only if a Sum Value is present in a MetricPoint, then the MetricPoint's +Inf Classic Bucket value MUST also appear in a Sample with a MetricName with the suffix "_count".

Classic Buckets MUST be sorted in number increasing order of "le", and the value of the "le" label MUST follow the rules for Canonical Numbers.

> All Classic Buckets must be represented, even the unpopulated ones that didn't measure any values.

Author: I'll remove it.

> All Classic Buckets must be represented, even the unpopulated ones that didn't measure any values.

Author: We discussed in DM that considering all the other things, maybe it's worth mentioning somehow.

Maybe say that implementation libraries should discourage users from frequently changing the Classic Bucket thresholds as it leads to possibly invalidating queries and making it impossible to execute some time aggregations.

hmm?

> All Classic Buckets must be represented, even the unpopulated ones that didn't measure any values.

Reviewer 2: I wouldn't imply too much for instrumentation libraries. 

How about adding a sentence like this in the section `##### Classic Buckets` (line 247 and following):

"Exposed thresholds SHOULD stay constant over time and between targets whose metrics are intended to be aggregated. A change of thresholds may prevent the affected histograms to be part of the same operation (e.g. an aggregation of different metrics or a rate calculation over time)."

> All Classic Buckets must be represented, even the unpopulated ones that didn't measure any values.

Author: It doesn't hurt to spell this out. (Like I did with the case where classic buckets are no longer used, then clear them).

Fixed in 03be949a52133c0d850eb1f12c34421df058d667

foo_created 1520430000.123
```

##### Histogram with exponential buckets

The MetricPoint's value MUST be a complex data type.

Histograms with exponential buckets use the integer native histogram data type.

> The integer native histogram data type is a JSON like structure with fields. There MUST NOT be any whitespace around fields.

Reviewer 2: It's not really JSON. Many language use vaguely this kind of structure. I wouldn't call out JSON because it creates the expectation that other JSON syntax rules also apply.

Reviewer 2: > There MUST NOT be any whitespace around fields.

Has it already been decided that OMv2 will stay whitespace intolerant?

It's one of my big big red flags with OMv1 that it doesn't tolerate whitespace like the classic Prometheus text format. If OMv2 will see the light and allow whitespace again, then of course we should allow whitespace here.

Author: I'd like to defer this to a separate PR.

Reviewer 2: I resolve this to mean "I have made my point". I understand that the actual decision has been deferred. (Should OMv2 decide to be whitespace tolerant (which I hope), then I would argue that the representation of native histograms should be whitespace tolerant, too.)

The integer native histogram data type MUST include the Count, Sum, Schema, Zero Threshold, Zero bucket value as the fields `count`, `sum`, `schema`, `zero_threshold`, `zero_count`.

If there are no negative exponential buckets, then the fields `negative_spans` and `negative_deltas` SHOULD be omitted.
If there are no positive exponential buckets, then the fields `positive_spans` and `positive_deltas` SHOULD be omitted.

If there are negative (and/or positive) exponential buckets then the fields `negative_spans`, `negative_deltas` (and/or `positive_spans`, `positive_deltas`) MUST be present in this order after the `zero_count` field.

Exponential bucket values MUST be ordered by their index, and their values MUST be placed in the `negative_deltas` (and/or `positive_deltas`) field using delta encoding, that is the first bucket value is written as is and the following values only as a delta relative to the previous value. For example bucket values 1, 5, 4, 4 will become 1, 4, -1, 0.

> To map the `negative_deltas` (and/or `positive_deltas`) back to their indices, the `negative_spans` (and/or `positive_spans`) field MUST be constructed in the following way: each span consists of a pair of numbers, an integer called offset and an non-negative integer called length. Only the first span in each list can have a negative offset. It defines the index of the first bucket in its corresponding `negative_deltas` (and/or `positive_deltas`). The length defines the number of consecutive buckets the bucket list starts with. The offsets of the following spans define the number of excluded (and thus unpopulated buckets). The lengths define the number of consecutive buckets in the list following the excluded buckets.

Reviewer 2: Capitalize after `:`: each → Each

If a Histogram MetricPoint has both classic and exponential buckets, the exponential buckets MUST come first and the created time MUST NOT be duplicated.

The order ensures that implementations can easily skip the classic buckets if the exponential buckets are preferred.

```openmetrics-add-eof
# TYPE acme_http_request_seconds histogram
# UNIT acme_http_request_seconds seconds
# HELP acme_http_request_seconds Latency histogram of all of ACME's HTTP requests.
acme_http_request_seconds{path="/api/v1",method="GET"} {count:2,sum:1.2e2,schema:0,zero_threshold:1e-4,zero_count:0,positive_spans:[1:2],positive_deltas:[1,0]}
acme_http_request_seconds_count{path="/api/v1",method="GET"} 2
acme_http_request_seconds_sum{path="/api/v1",method="GET"} 1.2e2
acme_http_request_seconds_buckets{path="/api/v1",method="GET",le="0.5"} 1
acme_http_request_seconds_buckets{path="/api/v1",method="GET",le="1"} 2
acme_http_request_seconds_buckets{path="/api/v1",method="GET",le="+Inf"} 2
acme_http_request_seconds_created{path="/api/v1",method="GET"} 1605281325.0
```

###### Exemplars

Exemplars without Labels MUST represent an empty LabelSet as {}.

An example of Exemplars showcasing several valid cases:

> The native histogram version of the histogram has multiple Exemplars.

Reviewer 2: But I see only one?

Author: There's two 
```
foo {count:10,sum:1.0,schema:0,zero_threshold:1e-4,zero_count:0,positive_spans:[0:2],positive_deltas:[5,0]} # {trace_id="shaZ8oxi"} 0.67 1520879607.789 # {trace_id="ookahn0M"} 1.2 1520879608.589
```

The MetricPoint's value MUST be a complex data type.

Histograms with native buckets use the integer native histogram data type.

The integer native histogram data type is a JSON like structure with fields. There MUST NOT be any whitespace around fields.
The integer native histogram data type MUST include the Count, Sum, Schema, Zero Threshold, Zero native bucket value as the fields `count`, `sum`, `schema`, `zero_threshold`, `zero_count`.

If a positive or negative native bucket is empty, it SHOULD be omitted. See later for a valid use case when to keep empty buckets.

If there are no negative native buckets, then the fields `negative_spans` and `negative_buckets` SHOULD be omitted.
If there are no positive native buckets, then the fields `positive_spans` and `positive_buckets` SHOULD be omitted.

If there are negative (and/or positive) native buckets then the fields `negative_spans`, `negative_buckets` (and/or `positive_spans`, `positive_buckets`) MUST be present in this order after the `zero_count` field.

Native bucket values MUST be ordered by their index, and their values MUST be placed in the `negative_buckets` (and/or `positive_buckets`) fields.

> To map the `negative_buckets` (and/or `positive_buckets`) back to their indices, the `negative_spans` (and/or `positive_spans`) field MUST be constructed in the following way: each span consists of a pair of numbers, an integer called offset and an non-negative integer called length. Only the first span in each list can have a negative offset. It defines the index of the first bucket in its corresponding `negative_buckets` (and/or `positive_buckets`). The length defines the number of consecutive buckets the bucket list starts with. The offsets of the following spans define the number of excluded (and thus unpopulated buckets). The lengths define the number of consecutive buckets in the list following the excluded buckets.

Reviewer 2: ```suggestion
To map the `negative_buckets` (and/or `positive_buckets`) back to their indices, the `negative_spans` (and/or `positive_spans`) field MUST be constructed in the following way: Each span consists of a pair of numbers, an integer called offset and an non-negative integer called length. Only the first span in each list can have a negative offset. It defines the index of the first bucket in its corresponding `negative_buckets` (and/or `positive_buckets`). The length defines the number of consecutive buckets the bucket list starts with. The offsets of the following spans define the number of excluded (and thus unpopulated buckets). The lengths define the number of consecutive buckets in the list following the excluded buckets.
```

foo_sum 324789.3 st@1520430000.123
```

> ##### Histogram with native buckets

Reviewer 3: ```suggestion
##### Histogram with Native Buckets
```

The MetricPoint's value MUST be a complex data type.

Histograms with Native Buckets MUST use the integer native histogram data type.

The integer native histogram data type is represented as structured data with fields. There MUST NOT be any whitespace around fields.

> The integer native histogram data type MUST include the Count, Sum, Schema, Zero Threshold, Zero Native Bucket value as the fields `count`, `sum`, `schema`, `zero_threshold`, `zero_count`, in this order.

Author: 6b8f369dbacfd30c818e10896714f36d2bc8554f

foo_sum 324789.3 ct@1520430000.123
```

##### Histogram with native buckets

The MetricPoint's value MUST be a complex data type.

Histograms with Native Buckets MUST use the integer native histogram data type.

The integer native histogram data type is represented as structured data with fields. There MUST NOT be any whitespace around fields.
The integer native histogram data type MUST include the Count, Sum, Schema, Zero Threshold, Zero Native Bucket value as the fields `count`, `sum`, `schema`, `zero_threshold`, `zero_count`, in this order.

If there are no negative Native Buckets, then the fields `negative_spans` and `negative_buckets` SHOULD be omitted.
If there are no positive Native Buckets, then the fields `positive_spans` and `positive_buckets` SHOULD be omitted.

If there are negative (and/or positive) Native Buckets, then the fields `negative_spans`, `negative_buckets` (and/or `positive_spans`, `positive_buckets`) MUST be present in this order after the `zero_count` field.

With the exception of the `sum` and `zero_threshold` field, all numbers MUST be integers and MUST NOT include dot '.' or exponent 'e'.

> Native Bucket values MUST be ordered by their index, and their values MUST be placed in the `negative_buckets` (and/or `positive_buckets`) fields.

Author: note to reviewers: we don't use delta encoding here, `negative_buckets` and `positive_buckets` hold the actual absolute non negative integer value.
